File: Dome/dome/rules/scanner.py
"""Scanner / automated tool detection rules."""
import re

# Known scanner / exploit-tool User-Agent substrings (lowercase)
SCANNER_UA = [
    "sqlmap", "nikto", "nmap", "masscan", "dirbuster", "dirb", "gobuster",
    "wfuzz", "ffuf", "burpsuite", "burp ", "owasp zap", "zaproxy",
    "acunetix", "nessus", "openvas", "metasploit", "msfconsole",
    "havij", "pangolin", "w3af", "skipfish", "arachni",
    "python-requests", "go-http-client", "libwww-perl", "lwp-trivial",
    "curl/", "wget/", "httpie", "nuclei", "dalfox", "commix",
]

# Paths that scanners probe (case-insensitive)
SCAN_PATHS = [
    r"(?i)/\.git/",
    r"(?i)/\.env",
    r"(?i)/wp-login\.php",
    r"(?i)/xmlrpc\.php",
    r"(?i)/admin/config",
    r"(?i)/phpmyadmin",
    r"(?i)/(etc|proc)/",
    r"(?i)/shell\.(php|asp|aspx|jsp)",
    r"(?i)/c99|r57|b374k",           # known webshell names
    r"(?i)/eval-stdin",
    r"(?i)/actuator/(env|heapdump|mappings|beans)",   # Spring Boot
    r"(?i)/__debugbar",
    r"(?i)/\.htaccess",
]

COMPILED_PATHS = [re.compile(p) for p in SCAN_PATHS]


def check_ua(ua: str) -> list[dict]:
    """Check User-Agent for scanner signatures."""
    hits = []
    ua_lower = ua.lower()
    for sig in SCANNER_UA:
        if sig in ua_lower:
            hits.append({
                "rule_id": "SCAN-001",
                "description": f"Known scanner User-Agent: {sig}",
                "category": "scanner",
            })
            break
    return hits


def check_path(path: str) -> list[dict]:
    """Check URL path for known probe/scanner patterns."""
    hits = []
    for pattern in COMPILED_PATHS:
        if pattern.search(path):
            hits.append({
                "rule_id": "SCAN-002",
                "description": f"Scanner probe path: {path}",
                "category": "scanner",
            })
            break
    return hits

File: Dome/dome-rs/src/engine/scanner.rs
/// Scanner / automated tool detection.
use once_cell::sync::Lazy;
use regex::Regex;
use super::Hit;

const SCORE_UA:   u32 = 15;
const SCORE_PATH: u32 = 10;

// Known scanner User-Agent substrings (lowercase)
static SCANNER_UAS: &[&str] = &[
    "sqlmap", "nikto", "nmap", "masscan", "dirbuster", "dirb", "gobuster",
    "wfuzz", "ffuf", "burpsuite", "burp ", "owasp zap", "zaproxy",
    "acunetix", "nessus", "openvas", "metasploit", "msfconsole",
    "havij", "pangolin", "w3af", "skipfish", "arachni",
    "python-requests", "go-http-client", "libwww-perl", "lwp-trivial",
    "nuclei", "dalfox", "commix", "hydra", "medusa",
    "zgrab", "xsser", "wpscan", "joomscan",
];

static SCAN_PATHS: Lazy<Vec<Regex>> = Lazy::new(|| {
    [
        r"(?i)/\.git/",
        r"(?i)/\.env$",
        r"(?i)/wp-login\.php",
        r"(?i)/xmlrpc\.php",
        r"(?i)/phpmyadmin",
        r"(?i)/(etc|proc)/",
        r"(?i)/shell\.(php|asp|aspx|jsp)",
        r"(?i)/(c99|r57|b374k|webshell)",
        r"(?i)/eval-stdin",
        r"(?i)/actuator/(env|heapdump|mappings|beans|dump)",
        r"(?i)/__debugbar",
        r"(?i)/\.htaccess",
        r"(?i)/\.svn/",
        r"(?i)/backup\.(zip|tar|gz|sql)",
        r"(?i)/config\.(php|yml|yaml|json|bak)",
        r"(?i)/wp-content/debug\.log",
        r"(?i)/server-status",
        r"(?i)/manager/html",   // Tomcat
        r"(?i)/solr/admin",
        r"(?i)/jmx-console",
        r"(?i)/console/login",  // JBoss
    ].iter().map(|p| Regex::new(p).unwrap()).collect()
});

pub fn check_ua(ua: &str) -> Vec<Hit> {
    let lower = ua.to_lowercase();
    for sig in SCANNER_UAS {
        if lower.contains(sig) {
            return vec![Hit {
                rule_id:     "SCAN-001",
                description: "Known scanner/exploit-tool User-Agent",
                category:    "scanner",
                score:       SCORE_UA,
            }];
        }
    }
    vec![]
}

pub fn check_path(path: &str) -> Vec<Hit> {
    for pat in SCAN_PATHS.iter() {
        if pat.is_match(path) {
            return vec![Hit {
                rule_id:     "SCAN-002",
                description: "Scanner probe path detected",
                category:    "scanner",
                score:       SCORE_PATH,
            }];
        }
    }
    vec![]
}
